sklearn/linear_model/_linear_loss.py

```diff
@@ -0,0 +1,362 @@
+
+        return w, intercept, raw_prediction
+
+    def loss(self, coef, X, y, sample_weight=None, alpha=0.0, n_threads=1):
+        """Compute the loss as sum over point-wise losses.
+
+        Parameters
+        ----------
+        coef : ndarray of shape (n_dof,) or (n_classes * n_dof,)
+            Coefficients of a linear model.
+        y : C/F-contiguous array of shape (n_samples,)
+            Observed, true target values.
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+        sample_weight : None or C/F-contiguous array of shape (n_samples,)
+            Sample weights.
+        alpha: float
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        loss : float
+            Sum of losses per sample plus penalty.
+        """
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss = self._loss.loss(
+            y_true=y,
+            raw_prediction=raw_prediction,
+            sample_weight=sample_weight,
+            n_threads=n_threads,
+        )
+        loss = loss.sum()
+
+        if w.ndim == 1:
+            return loss + 0.5 * alpha * (w @ w)
+        else:
+            return loss + 0.5 * alpha * squared_norm(w)
+
+    def loss_gradient(self, coef, X, y, sample_weight=None, alpha=0.0, n_threads=1):
+        """Computes the sum/average of loss and gradient.
+
+        Parameters
+        ----------
+        coef : ndarray of shape (n_dof,) or (n_classes * n_dof,)
+            Coefficients of a linear model.
+        y : C/F-contiguous array of shape (n_samples,)
+            Observed, true target values.
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data.
+        sample_weight : None or C/F-contiguous array of shape (n_samples,)
+            Sample weights.
+        alpha: float
+            L2 regularization strength
+        n_threads : int, default=1
+            Number of OpenMP threads to use.
+
+        Returns
+        -------
+        loss : float
+            Sum of losses per sample plus penalty.
+
+        gradient : ndarray of shape (n_dof,) or (n_classes * n_dof)
+             The gradient of the loss as ravelled array.
+        """
+        n_features, n_classes = X.shape[1], self._loss.n_classes
+        n_dof = n_features + self.fit_intercept
+        w, intercept, raw_prediction = self._w_intercept_raw(coef, X)
+
+        loss, gradient = self._loss.loss_gradient(
+            y_true=y,
+            raw_prediction=raw_prediction,
+            sample_weight=sample_weight,
+            n_threads=n_threads,
+        )
+        loss = loss.sum()
+
+        if not self._loss.is_multiclass:
+            loss += 0.5 * alpha * (w @ w)
+            grad = np.empty_like(coef, dtype=X.dtype)
+            grad[:n_features] = X.T @ gradient + alpha * w
```

Comment:
Unfortunately, `grad = X.T @ gradient` does not work if `fit_intercept=True`. This array would be one element too short.

Reply:
What about still allocating with `np.empty`, but then
```python
np.dot(X.T, gradient, out=grad[:n_features])
```

Reply:
yes that would work and the code will be even easier to adapt if we introduce gradient buffer memory reuse later (instead of calling `np.empty` at each iteration).

Reply:
I'd would postpone this improvement to a possible follow-up PR because with `np.dot(..., out=..)` one needs a case distinction for sparse X which leads to more redundant code.